src/lsm/zig_zag_merge.zig

```diff
@@ -0,0 +1,546 @@
+    comptime stream_probe: fn (context: *Context, stream_index: u32, key: Key) void,
+) type {
+    return struct {
+        const ZigZagMergeIterator = @This();
+        const BitSet = std.bit_set.IntegerBitSet(streams_max);
+
+        context: *Context,
+        streams_count: u32,
+        direction: Direction,
+        previous_key_popped: ?Key = null,
+
+        /// At least two scans are required for zig-zag merge.
+        pub fn init(
+            context: *Context,
+            streams_count: u32,
+            direction: Direction,
+        ) ZigZagMergeIterator {
+            assert(streams_count <= streams_max);
+            assert(streams_count > 1);
+
+            return .{
+                .context = context,
+                .streams_count = streams_count,
+                .direction = direction,
+            };
+        }
+
+        // Resets the iterator when the underlying streams are moved.
+        // It's not necessary for ZigZagMerge, but it follows the same API for all MergeIterators.
+        pub fn reset(it: *ZigZagMergeIterator) void {
+            _ = it;
+        }
+
+        pub fn pop(it: *ZigZagMergeIterator) error{Drained}!?Value {
+            while (try it.peek_key()) |key| {
+                const value = stream_pop(it.context, 0);
+                assert(key_from_value(&value) == key);
+                for (1..it.streams_count) |stream_index| {
+                    const value_other = stream_pop(it.context, @intCast(stream_index));
+                    assert(key_from_value(&value_other) == key);
+
+                    if (constants.verify) {
+                        // Differently from K-way merge, there's no precedence between streams
+                        // in Zig-Zag merge. It's assumed that all streams will produce the same
+                        // value during a key intersection.
+                        assert(stdx.equal_bytes(Value, &value, &value_other));
+                    }
+                }
+
+                if (it.previous_key_popped) |previous| {
+                    switch (std.math.order(previous, key)) {
+                        .lt => assert(it.direction == .ascending),
+                        // Duplicate values are not expected.
+                        .eq => unreachable,
+                        .gt => assert(it.direction == .descending),
+                    }
+                }
+                it.previous_key_popped = key;
+
+                return value;
+            }
+
+            return null;
+        }
+
+        fn peek_key(it: *ZigZagMergeIterator) error{Drained}!?Key {
+            assert(it.streams_count <= streams_max);
+            assert(it.streams_count > 1);
+
+            const key_min: Key = switch (it.direction) {
+                .ascending => 0,
+                .descending => std.math.maxInt(Key),
+            };
+
+            var drained: BitSet = BitSet.initEmpty();
+            var probe_key: Key = key_min;
```

Comment:
:thinking: If we persisted this on `ZigZagMergeIterator`, then we could verify that it never regresses, i.e. verify the correct implementation of `stream_probe` across iterations, by checking that `stream_peek`'s return value is after-or-equal-to the latest probe key. (e.g. Right now if `ScanTree` had a bug in its probe implementation, I think it would be possible for zig-zag to endlessly retry probe-drained-probe-drained and never advance.)